### ISSUE TITLE
DbUpdateException seems to have deleted original row when a duplicate insertion exception occurs

### ISSUE THREAD

**Issue by yashish** [dotnet/efcore#18337]
I'm perplexed when a duplicate insert exception occurs, _**I found the original row got deleted when the illegal insertion tried to occur**_ due to some bug in our code. Is this happening because of using the **DeleteBehavior.ClientSetNull** action?

Code below showing the exception  thrown and EF Core generated DbContext entity.

Affected Entity:
```C#
                entity.HasOne(d => d.DigitalCampaign)
                    .WithMany(p => p.CampaignOperativeProduct)
                    .HasForeignKey(d => d.DigitalCampaignId)
                    .OnDelete(DeleteBehavior.ClientSetNull)
                    .HasConstraintName("CampaignOperativeProduct DigitalCampaignId fk");
```
Exception:
```
There was DbUpdateException error saving the Campaign a088A000002CWrIQAW in OMS database. O1 SalesOrder: 30002231.
 Error: System.Data.SqlClient.SqlException (0x80131904): Cannot insert duplicate key row in object 'OMS.CampaignOperativeProduct' with unique index 'DigitalCampaignId and OperativeProductId together must be unique'. The duplicate key value is (2325, 11).
   at System.Data.SqlClient.SqlConnection.OnError(SqlException exception, Boolean breakConnection, Action`1 wrapCloseInAction)
   at System.Data.SqlClient.SqlInternalConnection.OnError(SqlException exception, Boolean breakConnection, Action`1 wrapCloseInAction)
   at System.Data.SqlClient.TdsParser.ThrowExceptionAndWarning(TdsParserStateObject stateObj, Boolean callerHasConnectionLock, Boolean asyncClose)
   at System.Data.SqlClient.TdsParser.TryRun(RunBehavior runBehavior, SqlCommand cmdHandler, SqlDataReader dataStream, BulkCopySimpleResultSet bulkCopyHandler, TdsParserStateObject stateObj, Boolean& dataReady)
   at System.Data.SqlClient.SqlDataReader.TryHasMoreRows(Boolean& moreRows)
   at System.Data.SqlClient.SqlDataReader.TryHasMoreResults(Boolean& moreResults)
   at System.Data.SqlClient.SqlDataReader.TryNextResult(Boolean& more)
   at System.Data.SqlClient.SqlDataReader.<>c__DisplayClass187_0.<NextResultAsync>b__1(Task t)
   at System.Data.SqlClient.SqlDataReader.InvokeRetryable[T](Func`2 moreFunc, TaskCompletionSource`1 source, IDisposable objectToDispose)
--- End of stack trace from previous location where exception was thrown ---
   at Microsoft.EntityFrameworkCore.Update.AffectedCountModificationCommandBatch.ConsumeAsync(RelationalDataReader reader, CancellationToken cancellationToken) ClientConnectionId:d3869a6f-d7ba-4ed4-bae7-1e5a2c3f2e6a
Error Number:2601,State:1,Class:14 
```

**Comment by ajcvickers** [comment]
@yashish It's hard to give a conclusive answer with just the code shown. Please post a small, runnable project/solution or complete code listing that demonstrates the behavior you are seeing.

**Comment by yashish** [comment]
@ajcvickers it's just this and a c# code with EF Core 2.2.4 that adds the entity. No transactions or triggers involved and not even soft deletes for any entity added in the system. The system we built was under test by the business with no access to the backend. So when I discovered that original records got deleted when an EF add operation failed, I was wondering if it was a configuration issue. But that doesn't seem to be the case either. Must be something I missed that the business did in their testing use case scenario. I'm closing the issue as I don't have any further information on this, but thank you for getting back.
